src/DataStructures/Mesh.hpp

```diff
@@ -0,0 +1,115 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class template Mesh.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+
+#include "DataStructures/Index.hpp"
+#include "Utilities/Requires.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+namespace Spectral {
+enum class Basis;
+enum class Quadrature;
+}  // namespace Spectral
+/// \endcond
+
+template <size_t Dim>
+class Mesh : public Index<Dim> {
+ public:
+  Mesh() noexcept = default;
+
+  /*!
+   * \brief Construct a computational grid with the same number of collocation
+   * points in each dimension.
+   *
+   * \param isotropic_extents The number of collocation points in each
+   * dimension
+   * \param basis The choice of spectral basis to compute the
+   * collocation points
+   * \param quadrature The choice of quadrature to compute
+   * the collocation points
+   */
+  Mesh(const size_t isotropic_extents, Spectral::Basis basis,
+       Spectral::Quadrature quadrature) noexcept
+      : Index<Dim>(isotropic_extents) {
+    bases_.fill(basis);
+    quadratures_.fill(quadrature);
+  }
+
+  /*!
+   * Construct a computational grid where each dimension can have a different
+   * number of collocation points.
+   *
+   * \param extents The number of collocation points per dimension
+   * \param basis The choice of spectral basis to compute the
+   * collocation points
+   * \param quadrature The choice of quadrature to compute
+   * the collocation points
+   */
+  Mesh(std::array<size_t, Dim> extents, Spectral::Basis basis,
+       Spectral::Quadrature quadrature) noexcept
+      : Index<Dim>(std::move(extents)) {
+    bases_.fill(basis);
+    quadratures_.fill(quadrature);
+  }
+
+  /*!
+   * Construct a computational grid where each dimension can have both a
+   * different number and placement of collocation points.
+   *
+   * \param extents The number of collocation points per dimension
+   * \param bases The choice of spectral bases to compute the
+   * collocation points per dimension
+   * \param quadratures The choice of quadratures to compute
+   * the collocation points per dimension
+   */
+  Mesh(std::array<size_t, Dim> extents, std::array<Spectral::Basis, Dim> bases,
+       std::array<Spectral::Quadrature, Dim> quadratures) noexcept
+      : Index<Dim>(std::move(extents)),
+        bases_(std::move(bases)),
+        quadratures_(std::move(quadratures)) {}
+
+  /*!
+   * \brief The basis chosen in each dimension of the grid.
+   */
+  std::array<Spectral::Basis, Dim> bases() const noexcept { return bases_; }
```

Comment:
Should this be returned by const reference instead of value? Same with quadratures